src/chess/bitboard.h

```diff
@@ -190,40 +190,43 @@ class Move {
  public:
   enum class Promotion : std::uint8_t { None, Queen, Rook, Bishop, Knight };
   Move() = default;
-  Move(BoardSquare from, BoardSquare to) : from_(from), to_(to) {}
+  Move(BoardSquare from, BoardSquare to)
+      : data_(to.as_int() + (from.as_int() << 6)) {}
   Move(BoardSquare from, BoardSquare to, Promotion promotion)
-      : from_(from), to_(to), promotion_(promotion) {}
+      : data_(to.as_int() + (from.as_int() << 6) +
+              (static_cast<uint8_t>(promotion) << 12)) {}
   Move(const std::string& str, bool black = false);
   Move(const char* str, bool black = false) : Move(std::string(str), black) {}
 
-  BoardSquare from() const { return from_; }
-  BoardSquare to() const { return to_; }
-  Promotion promotion() const { return promotion_; }
-  bool IsCastling() const { return castling_; }
-  void SetCastling() { castling_ = true; }
+  BoardSquare to() const { return BoardSquare(data_ & kToMask); }
+  BoardSquare from() const { return BoardSquare((data_ & kFromMask) >> 6); }
+  Promotion promotion() const { return Promotion((data_ & kPromoMask) >> 12); }
+  bool castling() const { return (data_ & kCastleMask) != 0; }
+  void SetCastling() { data_ |= kCastleMask; }
 
+  void SetTo(BoardSquare to) { data_ = (data_ & ~kToMask) | to.as_int(); }
+  void SetFrom(BoardSquare from) {
+    data_ = (data_ & ~kFromMask) | (from.as_int() << 6);
+  }
+  void SetPromotion(Promotion promotion) {
+    data_ = (data_ & ~kPromoMask) | (static_cast<uint8_t>(promotion) << 12);
+  }
   // 0 .. 16384, knight promotion and no promotion is the same.
   uint16_t as_packed_int() const;
 
   // 0 .. 1857, to use in neural networks.
   uint16_t as_nn_index() const;
 
-  bool operator==(const Move& other) const {
-    return from_ == other.from_ && to_ == other.to_ &&
-           promotion_ == other.promotion_;
-  }
+  bool operator==(const Move& other) const { return data_ == other.data_; }
```

Comment:
Any idea why the old equality didn't compare the castling bit?

Reply:
no, but on the previous version of this PR Crem said "Indeed it should be data_ == other.data_, castlings are expected to be different moves."